circle-mlir/circle-mlir/lib/pass/src/ops/BatchNormalizationInferenceModeOp.h

```diff
@@ -0,0 +1,164 @@
+    LLVM_DEBUG({ llvm::dbgs() << "BatchNormIM momentum: " << momentum_val << "\n"; });
+
+    // NOTE below (a bit complicated) subgraph will be shrunk to Mul-Add by constant folding
+
+    // 0/ broadcast scale, bias, mean, variance to match 1xCx1x1
+    //    only rank 4 Constant works
+    if (intype.getRank() == 4)
+    {
+      scale = CreateConstBroadcastChn(rewriter, input, scale, op_name + "/scale");
+      bias = CreateConstBroadcastChn(rewriter, input, bias, op_name + "/bias");
+      mean = CreateConstBroadcastChn(rewriter, input, mean, op_name + "/mean");
+      variance = CreateConstBroadcastChn(rewriter, input, variance, op_name + "/variance");
+    }
+
+    // 1/ from onnx-tensorflow
+    //   running_mean = mean * momentum + mean * (1 - momentum)
+    //   running_variance = variance * momentum + variance * (1 - momentum)
+    //   inputs = [x, running_mean, running_variance, bias, scale]
+
+    mlir::Value const_momentum = CreateConst(rewriter, mean, momentum_val, op_name + "/momentum");
+    mlir::Value const_1_momentum =
+      CreateConst(rewriter, mean, 1.0f - momentum_val, op_name + "/1_momentum");
+    assert(not const_momentum.getType().isa<mlir::NoneType>());
+    assert(not const_1_momentum.getType().isa<mlir::NoneType>());
+
+    // mul_mm = mean * momentum
+    mlir::Location mul_mm_loc = mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/mul_mm"));
+    mlir::Value mul_mm =
+      rewriter.create<MulOp>(mul_mm_loc, mean.getType(), mean, const_momentum, "NONE");
+    // mul_1_mm = mean * (1 - momentum)
+    mlir::Location mul_1_mm_loc = mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/mul_1_mm"));
+    mlir::Value mul_1_mm =
+      rewriter.create<MulOp>(mul_1_mm_loc, mean.getType(), mean, const_1_momentum, "NONE");
+    // running_mean = mean * momentum + mean * (1 - momentum)
+    mlir::Location running_mean_loc =
+      mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/running_mean"));
+    mlir::Value running_mean =
+      rewriter.create<AddOp>(running_mean_loc, mean.getType(), mul_mm, mul_1_mm, "NONE");
+
+    // mul_vm = variance * momentum
+    mlir::Location mul_vm_loc = mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/mul_vm"));
+    mlir::Value mul_vm =
+      rewriter.create<MulOp>(mul_vm_loc, mean.getType(), variance, const_momentum, "NONE");
+    // mul_1_vm = variance * (1 - momentum)
+    mlir::Location mul_1_vm_loc = mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/mul_1_vm"));
+    mlir::Value mul_1_vm =
+      rewriter.create<MulOp>(mul_1_vm_loc, mean.getType(), variance, const_1_momentum, "NONE");
+    // running_variance = variance * momentum + variance * (1 - momentum)
+    mlir::Location running_variance_loc =
+      mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/running_variance"));
+    mlir::Value running_variance =
+      rewriter.create<AddOp>(running_variance_loc, mean.getType(), mul_vm, mul_1_vm, "NONE");
```

Comment:
Ditto, it seems the `running_variance` is not used in the following codes.
We may comment out this lines.

```suggestion
    // only for completeness (not used)
    // // running_variance = variance * momentum + variance * (1 - momentum)
    // mlir::Location running_variance_loc =
    //   mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/running_variance"));
    // mlir::Value running_variance =
    //   rewriter.create<AddOp>(running_variance_loc, mean.getType(), mul_vm, mul_1_vm, "NONE");
```